### ISSUE TITLE
hashing breaks for custom objects with "Invalid value used as weak map key"

### ISSUE THREAD

**Issue by UnsungHero97** [immutable-js/immutable-js#1643]
I think the [latest changes to `Hash.js`](https://github.com/facebook/immutable-js/commit/a74700d22fa94b0bc1253fe166ee00487ca001d8#diff-8fbd09a266961312c5b404bb2a549e10) introduce a weird bug when hashing custom objects, more specifically, a class with `valueOf()` overridden. The error being thrown is `Uncaught TypeError: Invalid value used as weak map key`. 

It's strange because it doesn't happen until a certain point, a very specific point. It happens when invoking `.set()` on a `Map` on the 9th attempt. For the first 8 attempts, all is good. On the 9th attempt, error.

To reproduce, I ran this bit of code in the console at https://facebook.github.io/immutable-js.

```
class Example {
  constructor(id, key) { this.id = id; this.key = key; }
  valueOf() { return `${this.id}.${this.key}`; } 
  toString() { return `${this.id}.${this.key}`; }
}

const m = Immutable.Map().asMutable();
m.set(new Example('test', (new Date()).toISOString()), Math.random()); // 1, ok
...
m.set(new Example('test', (new Date()).toISOString()), Math.random()); // 8, ok
m.set(new Example('test', (new Date()).toISOString()), Math.random()); // 9, ERROR
```

This has worked for me for quite a while. It breaks with release `v4.0.0-rc.11`.

Here's a screenshot of the console:

![screen shot 2018-11-06 at 12 17 27 pm](https://user-images.githubusercontent.com/440299/48091279-19f01380-e1be-11e8-85af-fa607bf1916e.png)

**Comment by conartist6** [comment]
o_O

**Comment by jamiedavenport** [comment]
Is there a workaround for this or is anyone working on a fix?

**Comment by tyger** [comment]
Bumped into, what looks the same issue. Here is a codepen: https://codepen.io/pen/tour/welcome/start?editors=0010
Check console of the browser.

**Comment by kwek** [comment]
We also encountered this issue and the fix in https://github.com/adamshone/immutable-js/commit/209817ed0cd584379824ba5f028e3baf7ab48b5d seems to solve. Any chance this can be merged?

**Comment by futpib** [comment]
Using this as a workaround :man_facepalming: 	

```js
Date.prototype.hashCode = function () { 
	return Number(this);
};
```

**Comment by sureshjoshi** [comment]
I've had the same/similar bug manifest in a different way.

When calling `toSet()` on a `Map<string, T>`, in some conditions (that, for the life of me I couldn't figure out a small reproduction case for), I ran into the following error: `TypeError: Attempted to set a non-object key in a WeakMap`

It took me about 4 hours of debugging to figure out that it was occurring when `hashCode()` was being called on some of my Records. For some reason, calling `toSet()`, `toSetSeq()` (and maybe one or two more) threw exceptions... `toList()`, `toMap()`, `toArray()` all worked fine.

In any case, I merged @adamshone's PR - and it fixed my issue too. Just wanted to post this to save someone some grief when Googling down the road.

**Comment by abrcdf1023** [comment]
Have this problem also does anyone know how to fix this when use with Moment library?

**Comment by lukas1994** [comment]
just hit the same issue today
when is this PR getting merged?

**Comment by adamshone** [comment]
I fixed it over a year ago but I think this project has been abandoned. We ran with a fork for a while and eventually switched to https://github.com/immerjs/immer

¯\_(ツ)_/¯

**Comment by mattleonowicz** [comment]
I hit the same thing after adding couple of `Date`s to `Set`

**Comment by Zaidos** [comment]
> I hit the same thing after adding couple of `Date`s to `Set`

Was also able to reproduce this using luxon's `DateTime` types. Haven't been able to find a work around yet.

**Comment by lukas1994** [comment]
This is happening more frequently. Has anyone figured out a solution? Is immutable-js actually not maintained anymore?

**Comment by JoshuaToenyes** [comment]
> This is happening more frequently. Has anyone figured out a solution? Is immutable-js actually not maintained anymore?

@lukas1994 there hasn't been a release since Oct 30, 2018, which is unfortunate. We're using [immer](https://github.com/immerjs/immer) now, which is fantastic and actively developed and maintained.

**Comment by lukas1994** [comment]
Thanks! Gonna be a big refactor but hopefully worth it :)
Do you have any advice for someone moving from immutable-js to immer? Is there anything I should be careful with?

**Comment by JoshuaToenyes** [comment]
Gosh, that's definitely worth a blog post!

Honestly, I've found immer a lot easier to work with, especially with Typescript (typing with Immutable, specifically Records got really tricking in some cases). I don't have a list of specific things to watch out for off the top of my head. But if you're interested in reading up on a larger project that refactored from Immutable to immer, checkout out [Slate](https://github.com/ianstormtaylor/slate): https://github.com/ianstormtaylor/slate/issues/2190, https://github.com/ianstormtaylor/slate/issues/2345, and their milestone for tracking the process https://github.com/ianstormtaylor/slate/milestone/3.

**Comment by lukas1994** [comment]
Cool will check those out. Thanks :)

**Comment by conartist6** [comment]
I'm not as familiar with Immer, but it doesn't look like it has the same functional APIs. If you need functional APIs I recommend [iter-tools](https://github.com/iter-tools/iter-tools) (which I maintain). It works with Maps and Sets (and any iterables), and has many of the methods you'd be familiar with from the `Immutable.Seq` type.

**Comment by JoshuaToenyes** [comment]
@conartist6 @lukas1994 - yes, immer has a dramatically different API. It's focused _only_ on immutability, basically, whereas Immutable is a full-featured data structure library.

Thanks for mentioning iter-tools. I haven't seen your lib before, but I'll check it out!

**Comment by sureshjoshi** [comment]
@lukas1994  For what it's worth, I also switched to Immer from Immutable. It was mildly tedious, but pretty easy and only took an hour or two for a mid-sized project. I'm using Typescript, so it was all compilation errors, zero runtime errors.

Unfortunately, Immutable (which is pretty awesome) ends up being a very leaky abstraction, so it was littered everywhere. Had I thought about this at the time, I could have hidden it away a bit more, but removing all those calls littered everywhere was the bulk of my time.

I decided not to make that same mistake with Immer. At lot of immer examples also show something like redux, which I don't use. My apps are typically a Clean Arch style, so they have underlying domain entities - which are the things most important things that I want to be immutable. So, I created a BaseEntity class which is the only file that contains references to ImmerJS, and it itself is only a constructor and two methods.

I haven't touched this code in a while, since it's worked - so there might be better ways to do this with the latest immer.

```typescript
export abstract class BaseEntity implements Entity {
    public readonly id: string;

    constructor(props: Partial<ConstructorType<BaseEntity>> = {}) {
        const { id = "" } = props;
        this.id = id;
        (this as any)[immerable] = true;
    }

    public set<K extends keyof this>(key: K, value: this[K]): this {
        const result = produce(this, (draft) => {
            draft[key] = value;
        });
        return result;
    }

    public copyOf(obj: Partial<this>): this {
        const result = produce(this, (draft) => {
            for (const [key, value] of Object.entries(obj)) {
                draft[key] = value ?? this[key];
            }
        });
        return result;
    }
```

And it gets used like:

```typescript
const foo = new Foo({
    value1,
    value2,
    value3,
    value4,
    value5,
    value6,
});

const newFoo = foo.set("myKey1", newValue1);

const newerFoo = foo
    .set("myKey1", newValue1)
    .set("myKey2", newValue2)
    .set("myKey3", newValue3);

const newestFoo = foo.copyOf({
    myKey1: newValue1,
    myKey2: newValue2,
    myKey3: newValue3,
});
```

Not perfect, but has worked for me for a year.

**Comment by conartist6** [comment]
I've incorporated this fix into https://github.com/immutable-js-oss/immutable-js/pull/17. We hope to be merging the community maintained fork back into immutable before we release. Otherwise we'll release `immutable-oss@4.0.0`.

**Comment by jdeniau** [comment]
Will be fixed by #1833 via https://github.com/immutable-js/immutable-js/pull/1833/commits/45196e3c7c793e64daa134b72bb89b529f6d8318
